packages/ses/src/module-link.js

```diff
@@ -35,30 +37,96 @@ export const link = (
   // compartment is in context: the module record may be in another
   // compartment, denoted by moduleRecord.compartment.
   // eslint-disable-next-line no-use-before-define
-  return instantiate(compartmentPrivateFields, moduleAliases, moduleRecord);
+  return instantiate(
+    compartmentPrivateFields,
+    moduleAnalyses,
+    moduleAliases,
+    moduleRecord,
+  );
+};
+
+const validateStaticModuleRecord = (staticModuleRecord, moduleAnalyses) => {
+  const isObject = typeof staticModuleRecord === 'object';
+  if (!isObject) {
+    throw new TypeError(
+      `importHook must return a StaticModuleRecord, got ${staticModuleRecord}`,
+    );
+  }
+
+  const hasAnalysis = moduleAnalyses.has(staticModuleRecord);
+  const hasImports = Array.isArray(staticModuleRecord.imports);
+  const isThirdParty =
+    hasImports && typeof staticModuleRecord.execute === 'function';
+
+  if (!hasAnalysis && !isThirdParty) {
+    if (hasImports) {
+      throw new TypeError(
+        `importHook must return a StaticModuleRecord constructed within the same Compartment and Realm`,
```

Comment:
Hard to see how you get from these conditions to this conclusion. Worth an extended comment.